File: task2/ruby/main.rb
require 'set' # модуль для множеств
# Проверяем валидность email адреса
def check_legality(address)
  user = address.split('@').first# Выделяем часть до @ (имя пользователя)
  domN = address.split('@').second
  return false if not domN.include?(".")
  return false if user.length < 6 || user.length > 30# Проверяем длину имени (6-30 символов)
  return false if user.chars.any? { |c| "&=+<>,_'-".include?(c) }# Проверяем на наличие запрещенных символов
  # Проверяем точки в начале/конце и двойные точки
  return false if user.start_with?('.') || user.end_with?('.')
  return false if user.include?('..')
  
  true # Если все проверки пройдены
end
# Нормализуем email (удаляем лишнее)
def sanitize_name(address)
  user, domain = address.split('@', 2)# Разбиваем на имя и домен
  user = user.split('*').first.gsub('.', '')# Удаляем все после звездочки и все точки
  "#{user}@#{domain}"# Собираем emil обратно
end
# Основная программа
print "Введите количество адресов: "
n = gets.to_i # Читаем количество адресов
adrs = Set.new # Создаем множество для хранения уникальных адресов
n.times do
  # Читаем и чистим ввод
  raw_adr = gets.chomp.strip
  # Если адрес валидный
  if check_legality(raw_adr)
    # Нормализуем его
    adr = sanitize_name(raw_adr)
    # Добавляем в множество (автоматическая уникальность)
    adrs.add(adr)
  end
end
puts adrs.size

File: task2/php/main.php
<?php
function checkLegality($address) {
    // Проверяем, что адрес содержит '@'
    if (!str_contains($address, '@')) {
        return false;
    }
    // Извлекаем имя пользователя
    list($user, $domain) = explode('@', $address, 2);
    if(!str_contains($domain,'.')) {return false;}
    // Длина имени пользователя должна быть от 6 до 30 символов
    if (strlen($user) < 6 || strlen($user) > 30) {
        return false;
    }
    // Проверяем наличие недопустимых символов в имени пользователя
    $invalidChars = "&=+<>,_'-";
    for ($i = 0; $i < strlen($user); $i++) {
        if (str_contains($invalidChars, $user[$i])) {
            return false;
        }
    }
    // Имя пользователя не должно начинаться или заканчиваться точкой, а также содержать две точки подряд
    if (str_starts_with($user, '.') || str_ends_with($user, '.')) { return false; }
    if (str_contains($user, '..')) { return false; }
    return true;
}
function sanitizeName($address) {
    // Разбиваем адрес на имя пользователя и домен
    list($user, $domain) = explode('@', $address, 2);
    // Обрезаем часть после символа '*', затем удаляем все точки
    $user = explode('*', $user, 2)[0];
    $user = str_replace('.', '', $user);  
    return $user . '@' . $domain;
}

$a = (int) readline("Введите количество адресов: ");
$adrs = array();
for ($n = 0; $n < $a; $n++) {
    $rawAdr = trim(readline("Введите адрес: "));
    // Сначала проверяем исходный адрес на корректность
    if (checkLegality($rawAdr)) {
        // Если адрес корректен, выполняем санитайзинг
        $adr = sanitizeName($rawAdr);
        // Добавляем адрес, если он ещё не содержится в массиве
        if (!in_array($adr, $adrs)) {
            $adrs[] = $adr;
        }
    }
}
print(count($adrs) . "\n");?>